zebrad/src/components/mempool/storage/eviction_list.rs

```diff
@@ -0,0 +1,138 @@
+//! [`EvictionList`] represents the transaction eviction list with
+//! efficient operations.
+use std::{
+    collections::{HashMap, VecDeque},
+    time::{Duration, Instant},
+};
+
+use zebra_chain::transaction;
+
+/// An eviction list that allows to efficiently add entries, get entries,
+/// and remove older entries in the order they were inserted.
+pub struct EvictionList {
+    // Maps each TXID in the list to the most recent instant they were added.
+    unique_entries: HashMap<transaction::Hash, Instant>,
+    // The entries in the order they were inserted.
+    // This can be larger than `unique_entries` if a same txid is added
+    // multiple times. Its instant will be overwritten in
+    // `unique_entries` but all entries will kept in `unique_entries`.
+    ordered_entries: VecDeque<(transaction::Hash, Instant)>,
+    // The maximum size of `unique_entries`.
+    max_size: usize,
+    /// The mempool transaction eviction age limit.
+    /// Same as [`Config::eviction_memory_time`].
+    eviction_memory_time: Duration,
+}
+
+impl EvictionList {
+    /// Create a new [`EvictionList`] with the given maximum size and
+    /// eviction time.
+    pub fn new(max_size: usize, eviction_memory_time: Duration) -> Self {
+        Self {
+            unique_entries: Default::default(),
+            ordered_entries: Default::default(),
+            max_size,
+            eviction_memory_time,
+        }
+    }
+
+    /// Inserts a TXID in the list, keeping track of the time it was inserted.
+    ///
+    /// If the TXID is already in the list, its insertion time will be updated.
+    ///
+    /// All entries older than [`EvictionList::eviction_time`] will be removed.
```

Comment:
Nitpick: missed docs in field rename?

```suggestion
    /// All entries older than [`EvictionList::eviction_memory_time`] will be removed.
```